ENH: Adds cross2d to linalg (with Array API compatibility)

bmwoodruff opened the pull request:
This PR stems from the community meeting on Jun 5, 2024. This implements a new function cross2d that accepts (a stack of) 2-element vectors, in the same manner that cross currently does. Both np.cross2d and np.linalg.cross2d are added.
The np.linalg.cross2d is API compatible.
This PR closes #13718 and closes #26620.
Units tests for cross2d are included.
A new test_ValueError is added for cross.
Updated doc links for both funtions.
Added examples for cross2d. Cleaned up whitespace on both functions.

This involved mostly replicating the content related `cross` and adapting it appropriately for `cross2d`. At the community meeting, there was discussion about whether or not to make it only available in `linalg`, or also in `core`. I made it available in both, which we can easily change. 

I was worried that I would need more help, but things seemed to be pretty straight forward. I think I've got all the proper things in place. I'll put a message up on slack and add a release note once I've gotten some feedback. Thanks for encouraging me to give this a try. It was very enjoyable to see how it all fits together.

bmwoodruff committed:
ENH: Adds cross2d to core and linalg (with API compatibility)

This PR stems from the community meeting on Jun 5, 2024.
This implements a new function cross2d that accepts (a stack of)
2-element vectors, in the same manner that cross currently does.
Both np.cross2d and np.linalg.cross2d are added.
The np.linalg.cross2d is API compatible.
This PR closes #13718 and closes #26620.
Units tests for cross2d are included.
A new test_ValueError is added for cross.
Updated doc links for both funtions.
Added examples for cross2d. Cleaned up whitespace on both functions.

bmwoodruff committed:
Added missing white space to an operator.

bmwoodruff commented:
I fixed the linter issue and added `python tools/linter.py` to my workflow (reread the testing page again, my bad). I'm guessing the freebsd_test failure is something I can ignore for now.

mtsokol commented:
I think we should also update the deprecation warning in `cross` https://github.com/numpy/numpy/blob/8d23615d78b1e1b241a713bb38c7c826c37fe38f/numpy/_core/numeric.py#L1649-L1653 and point users to the `cross2d` function.

bmwoodruff commented:
I'm trying to work out how to phrase it.  Mixed arrays (one 2d and one 3d) are completely deprecated with no alternative. Since `cross` works with 3d, and `cross2d` works with 2d, then how about something like this:
```
    if a.shape[-1] == 2 or b.shape[-1] == 2:
        # Deprecated in NumPy 2.0, 2023-09-26
        warnings.warn(
            "Mixed arrays with 2-dimensional vectors are deprecated. "
            "Use cross for arrays of 3-dimensional vectors. "
            "Use cross2d for arrays of 2-dimensional vectors. "
            "(deprecated in NumPy 2.0)",
            DeprecationWarning, stacklevel=2
        )
```
I could just add a bit to the old message, but I like the one above better.
```
    if a.shape[-1] == 2 or b.shape[-1] == 2:
        # Deprecated in NumPy 2.0, 2023-09-26
        warnings.warn(
            "Arrays of 2-dimensional vectors are deprecated. "
            "Use arrays of 3-dimensional vectors instead, or "
            "use cross2d for arrays of only 2-dimensional vectors. "
            "(deprecated in NumPy 2.0)",
            DeprecationWarning, stacklevel=2
        )
```
Thoughts?

I'm planning to add examples to the docstrings that show the similarities (and differences, if any) between `cross2d` and `det`.

bmwoodruff commented:
I updated the examples in `cross` that had 2d vectors in them, showing how to use pad to add a z-component.  I'll push those changes up soon, but I have a question.  While making revisions, do we want the online checks to happen with each and every commit (eating up online resources) or should I [skip ci] for some of the intermediate commits?  As this is a work in progress, I'd love to know what the regular workflow is. I've updated the deprecation message, and cleaned up the examples in cross, but haven't pushed them yet.

bmwoodruff committed:
ENH: Adds cross2d to core and linalg (with API compatibility)

This PR stems from the community meeting on Jun 5, 2024.
This implements a new function cross2d that accepts (a stack of)
2-element vectors, in the same manner that cross currently does.
Both np.cross2d and np.linalg.cross2d are added.
The function np.linalg.cross2d is Array API compatible.
This PR closes numpy#13718 and closes numpy#26620.
Units tests for cross2d are included.
A new test_ValueError is added for cross.
Updated doc links for both funtions and linked to .rst.
Added examples for cross2d. Cleaned up whitespace on both functions.
Adjusted examples in cross to show how to avoid deprecation warning.
Adds an example to compare cross2d with (floating point) det.
Includes a release note.

bmwoodruff commented:
I finished adding examples that seem relevant. I updated the deprecation warning, and adjusted the current existing 2d examples in cross to show someone how to avoid the deprecation. I pushed the changes, and if all but the `freebsd_test` pass, I'll add a note to the mailing list. Happy weekend all. 

Another question I have (please help train me in good practice) is "How much should I be including in subsequent commit messages?" I'm assuming you will squash the entire commit trail when this is all done, so is it simpler to just have a properly formatted first commit message, followed by very brief messages, or is it simpler to have a [properly formatted commit message](https://numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message) with every subsequent commit/push (delete all commit messages when squashing, but the last)? You'll see my second commit was a brief message, while my third was properly formatted. Does it matter?  Do you have a preference?

bmwoodruff committed:
Changed det to linalg.det

bmwoodruff commented:
The [circleci warning](https://app.circleci.com/pipelines/github/numpy/numpy/27022/workflows/ed5c1dd3-233e-45b9-ab40-19b106b0144b/jobs/40533?invite=true#step-108-13808_150) about `det` makes perfect sense.  I'm surprised it didn't show up with `spin docs --clean` when I ran the tests on my machine. Odd. It's fixed now.

mattip commented:
I missed the meeting. It seems awkward to me to have two similar functions with different APIs. Can we somehow get away with only one `cross2d`?

rgommers reviewed (commented):
Thanks for working on this @bmwoodruff! 

We said in the meeting that we did not want this function in the main namespace; if it made sense to add it then only in `numpy.linalg`. That also adheres to the design rule of having every function available only in a single namespace, which we spent a lot of effort on enforcing/improving in NumPy 2.0.

The thing I'd still like to see is the one-liner which we discussed - what is it? I couldn't work it out in the last call, but the claim was that `np.cross` with size-2 arrays was a one-liner with `det`. Adding `np.linalg.cross2d` is premised, I think, on that one-liner being too inconvenient to let users write by hand.

bmwoodruff commented:
Thanks for the feedback. I'm not aware of all the design rules that have shaped NumPy over the last 4 years (expected). I did spend a bit of time over the weekend reading about the newer API standards, but I'm also still learning the jargon (it's slowly coming). I remembered hearing that just `linalg.cross2d` was preferred (not required, but you all probably knew that). I can definitely remove it. If the current version of myself were at the meeting last week, I'd probably take away a lot more meaning. 

What I did was to make `numpy2d` and `numpy.lingal.2d` mimic the corresponding functions, in terms of inputs, functionality, and returns. I was going to simplify things to use only one function, but noticed some differences. Here's a few observations: 

- Using `np.cross2d([1, 2], [3,4])` works while `np.linalg.cross2d([1, 2], [3,4])` does not, you must use `np.linalg.cross2d(np.array([1, 2]), np.array([3,4]))`. I'm guessing this is an API standard issue. If this is a 'user' error, then perhaps [the examples](https://numpy.org/devdocs/reference/generated/numpy.cross.html#numpy.cross) from `np.cross` need to be updated to meet this standard too. This could become a docs update issue (make the docstring examples API compatible). 

- I don't believe we can get similar functionality from a `linalg.cross2d` that we can get with `linalg.det`.  This is because `linalg.det` always returns a float. You cannot get a return value of an integer from integer inputs, etc. This could be a deal breaker for some people. For a general `det` function, returning floats makes sense.  For small matrices of integers, it may not be an acceptable solution. I have students who don't like to use NumPy precisely because of this issue.

> ... Adding `np.linalg.cross2d` is premised, I think, on that one-liner being too inconvenient to let users write by hand.

I did create an example in the new `cross2d` function to illustrate how to stack two arrays and then use `linalg.det`. I spent a few hours musing over how to utilize the axis options from `cross2d` in a "nice" way, and decided this was enough. Users who use different axes could learn to use the `moveaxis` command?
```
    The scalar cross product of 2-dimensional vectors is the determinant of
    a 2 by 2 matrix whose rows (or columns) are those vectors. Here we
    stack an array of vectors to illustrate the similarlity. Note the subtle
    difference where `det` returns a float even with integer inputs.

    >>> u = [[1, 2], [3, 4], [-1, 5]]
    >>> v = [[5, 6], [7, 8], [-2, 9]]
    >>> np.cross2d(u, v)
    array([-4, -4,  1])
    >>> stacked_array = np.stack((u, v), axis=1)
    >>> np.linalg.det(stacked_array)
    array([-4., -4.,  1.])

```
So a simple one-liner for default axes could be (haven't check fully if broadcasting would work with this). Hopefully this sparks some ideas of a better one-liner.
```
np.linalg.det(np.stack((u, v), axis=1))
```
Losing extra axis options is annoying, but for me the different return types seems like a huge hurdle.

Because of the type issues, I kept both functions there and was hoping for discussion to get more feedback and direction.

rgommers commented:
> I don't believe we can get similar functionality from a `linalg.cross2d` that we can get with `linalg.det`.  This is because `linalg.det` always returns a float.

Good point. Indeed, that doesn't seem desirable.

> Using `np.cross2d([1, 2], [3,4])` works while `np.linalg.cross2d([1, 2], [3,4])` does not

That looks like a bug in the `linalg.cross` implementation. It's expected that it takes array-like inputs, and it is documented as doing that. But the implementation is failing to call `np.asarray` on its inputs it looks like. We should get that fixed, and add a test. Would you be interested in submitting a separate PR to fix that up?

> Because of the type issues, I kept both functions there and was hoping for discussion to get more feedback and direction.

We really don't want the one in the main namespace; @seberg and I both commented on that in the call last week. That there's two versions of `cross` now is for historical reasons; for new functions we always want to add them only in one place.

seberg commented:
> That there's two versions of cross now is for historical reasons; for new functions we always want to add them only in one place.

I agree that there seems no reason to alias this outside of `np.linalg`, it's niche and `np.cross` can refer to it.  I don't really have an issue about aliases in topical namespaces when something should be in the main one, though.
(In other words, I wouldn't mind if we had `np.linalg.cross` even though we have to keep `np.cross`; but dunno how others feel.)

rgommers commented:
> I don't really have an issue about aliases in topical namespaces when something should be in the main one, though.

No design rule is absolute of course, we can always decide to make exceptions when they seem appropriate. But let's cross that bridge if and when we get to it, rather than discuss hypotheticals here? It's pretty rare for us to add new functions in the main namespace.

seberg commented:
I didn't want to discuss it, just clarify that my opinion may not be as strict as it may seem when it comes to general principles.

bmwoodruff commented:
> Would you be interested in submitting a separate PR to fix that up?

Yes! All the training I can get is great. I'll submit it as a draft to start with. I think I know what needs to be done.

> We really don't want the one in the main namespace;

Knowing that the input type issue was a bug, and not part of the API functionality, helps me see how to make just one function `np.linalg.cross2d`. That should be a simple fix and ready before the triage meeting.

bmwoodruff committed:
Removed cross2d and updated linalg.cross2d

bmwoodruff committed:
Bug: fixed lint issue in test_linalg.py

DWesl commented on numpy/linalg/_linalg.py:
```diff
@@ -3174,6 +3174,142 @@ def cross(x1, x2, /, *, axis=-1):
+    x2_1 = x2[..., 1]
+
+    # x1_0 * x2_1 - x1_1 * x2_0
+    multiply(x1_0, x2_1, out=cp)
```
Is there some particular reason you're not letting this line create/allocate `cp`?  At first glance it would appear to have the same result.

bmwoodruff replied:
I wanted to follow the code in `cross` as closely as possible, so that errors in one could be traced to errors in the others. For a really high dimensional array (with appropriate dim 2 on the correct axis), I wasn't sure how much to simplify the existing code from `cross`, so I kept it all. Here's the relevant bit from the creation of `cp` in `cross`. 

https://github.com/numpy/numpy/blob/ee73d0d9e75cae2ad09f328ce34ffd2fe9498dab/numpy/_core/numeric.py#L1655-L1687

I just trimmed off the parts that had no relevance, and swapped things to `x1` and `x2` (different naming conventions in `linalg`). I can see that defining `cp` this way in `cross` is needed because it checks `cp.ndim` first, a check not needed in `cross2d`. 
 
If I'm understanding correctly, you are suggesting the following:
- Remove L3292-L3295 entirely (as well as the import statements for `broadcast` and `promote_type`)
- Change L3308 to be `cp = x1_1 * x2_0`.  

Sounds completely reasonable.

bmwoodruff replied:
Looks like `promote_types` may need to stay, unless it's not needed to recast `x1` and `x2` as `dtype`. That still removes `broadcast` and creating an empty `cp`.

bmwoodruff replied:
Looks like none of this is needed at all, with the one-liner update from @seberg.

bmwoodruff commented:
At triage, @seberg suggested the following one liner:
```
cross2d = arr1[..., 0] * arr2[..., 1] - arr1[..., 1] * arr2[..., 0]
```
I'll test if this works (with a move axis before things), and then report what I find.

bmwoodruff committed:
ENH: Replaced computation for cross2d with one-liner

bmwoodruff commented:
The one-liner passed all tests.  I think the only thing left may be to update the docstring of `cross` to show how to use this one-liner.

bmwoodruff committed:
BUG: added asanyarray to end of cross2d, cleaned up rest.

I updated the docstrings.
I removed broadcast and promote_types from the import as not needed.
I added asanyarray right before the return to catch single values.
I swapped to asanyarray at the start to match other new functions.

bmwoodruff commented:
I'm pretty sure it's ready to go at this point and others have reviewed it. Please add @seberg as a coauthor. Thanks for the one-liner and helpful suggestions.

DWesl commented on numpy/linalg/_linalg.py:
```diff
@@ -3174,6 +3174,128 @@ def cross(x1, x2, /, *, axis=-1):
+    x1 = asanyarray(x1)
+    x2 = asanyarray(x2)
+
+    axis1, axis2 = (axis,) * 2
```
Would
```suggestion
    axis1 = axis2 = axis
```
work here?  I'm not sure if that would be in line with the style guide or not

bmwoodruff replied:
It would work, or just two lines. I got the code I used by just copying what was in cross (3d). Not sure what the better `style` option is, and would happily let a developer change it as they see fit.

DWesl commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -2315,6 +2315,56 @@ def test_cross():
+
+        assert_equal(actual, expected)
+
+    def test_2x2_anticommutativity(self):
```
[It looks like NumPy already uses](https://github.com/numpy/numpy/blob/86c7f4b3dbdab809ee03b711e62f22206f95b1d7/requirements/test_requirements.txt#L6) [Hypothesis](https://hypothesis.works) for testing, and this would be a decent place for it.

I would suggest [insisting](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.floats) the vector components be less than the square root of half the maximum representable value, or allowing the results to both be NaN if they are not negatives of each other.

I'm not sure if `assert_equal(la.cross2d(u, u), 0)` should be its own check or not.

bmwoodruff replied:
Haven't touched Hypothesis, and I probably don't have the bandwidth to pick that up right now.  I would happily take a suggested fix, and add you in a a coauthor if you want to clean up the test accordingly.  In this section, I modified the cross (3d) stuff again, and then added a few other things that I noticed were causing problems.

DWesl replied:
I opened a PR as bmwoodruff/numpy#1, or you could accept the suggestion in the GitHub file review interface and update the imports yourself.

bmwoodruff replied:
I've been waiting to hear the status on whether or not we'll be adding a cross2d to the code base.  I saw the PR on my repo, but haven't merged it in yet because I haven't taken the time to review it and know what it is doing. If/when it's decided to add in a cross2d, then I'll make the bandwidth to review the PR and merge it in. 

For now, I'm waiting. If there is time at the next triage meeting, I hope to discuss this and merge things.  Perhaps the suggested one-liner by @seberg is enough to remove any need for a cross2d.

Maybe the right direction to head is to get feedback from those using the deprecated 2D version on whether or not the suggested 1 line change is sufficient.

DWesl replied:
That sounds sensible

DWesl commented on numpy/linalg/tests/test_linalg.py:
```diff
@@ -2315,6 +2315,56 @@ def test_cross():
+
+        assert_equal(actual, expected)
+
+    def test_2x2_anticommutativity(self):
+        u = [1, 2]
+        v = [3, 4]
+        z = -2
+        cp = np.linalg.cross2d(u, v)
+        assert_equal(cp, z)
+        cp = np.linalg.cross2d(v, u)
+        assert_equal(cp, -z)
```
```suggestion
    @given(
        st.lists(
            st.one_of(st.integers(-10**9, 10**9), st.floats()), min_length=4, max_length=4
        )
    )
    def test_2x2_anticommutativity(self, concatenated_vectors):
        u = concatenated_vectors[:2]
        v = concatenated_vectors[2:]
        forward = np.linalg.cross2d(u, v)
        reverse = np.linalg.cross2d(v, u)
        assert_equal(forward, -reverse)
```
This assumes you've done `from hypothesis import given, strategies as st` somewhere above.
(current spellings of those imports [vary](https://github.com/search?q=repo%3Anumpy%2Fnumpy+%28%22import+hypothesis%22+OR+%22from+hypothesis%22%29&type=code), but this looks decent).
It seems `assert_equal` already handles nans, which simplifies the check.

I'm pretty sure this will work with `assert_equal` rather than `assert_allclose`, `assert_array_almost_equal`, or `assert forward == pytest.approx(-reverse)`, but I keep running into floating-point corner cases, so it's best you're aware of the options before the tests start failing.

The integer range provided should avoid overflow on 64-bit default integers: +/- 32,765 should do the same for 32-bit integers.

bmwoodruff commented:
With the one-liner addition to the cross page, I think we can close this PR until enough interest sparks from somewhere else to justify adding a new function cross2d. We can discuss this at triage review, or just close it for now and reopen it if/when popular demand calls for it (or firmly decide against it, and just point people to the one-liner). It was a useful exercise for me to create the function and integrate it in, if nothing else.

I vote for closing this without merging it in.

mattip commented:
At the triage meeting we decided to close this. If there are dissenters they should open a new issue to discuss it.

burnpanck commented:
Sad to hear that. **There is no good one-line implementation** of the functionality provided by the `np.cross2d` function. For starters, the proposed one-liner requires you to spell the arguments twice, which is not acceptable if they are expressions rather than variables. Adding-in support for the `axis=` argument easily makes this a 5-liner, and the code now is completely opaque to the geometric significance of the operation. Sure, I could export those 5 lines into a function, and call that function `cross2d`. Heck, why don't we include such a function into numpy?
